File: solutions/go/protein-translation/2/protein_translation.go
package protein

import (
	"bytes"
	"errors"
	"strings"
)

var (
	ErrStop        = errors.New("stop codon detected")
	ErrInvalidBase = errors.New("faulty protein")
)

const CodonLength int = 3

//BenchmarkProtein-8        557952              1935 ns/op             848 B/op         33 allocs/op
func FromRNA(rna string) (proteins []string, e error) {
	codonBuilder := strings.Builder{}
	buf := bytes.Buffer{}
	buf.WriteString(rna)
	codonSlice := buf.Next(CodonLength)
	for lengthSlice := len(codonSlice); lengthSlice > 0; {
		codonBuilder.Write(codonSlice)
		codon, err := FromCodon(codonBuilder.String())
		if err == nil {
			codonBuilder.Reset()
			proteins = append(proteins, codon)
		} else if err == ErrStop {
			return proteins, nil
		} else {
			return nil, ErrInvalidBase
		}
		codonSlice = buf.Next(CodonLength)
		lengthSlice = len(codonSlice)
	}
	return
}

//BenchmarkCodon-8         7037148               173.8 ns/op             0 B/op          0 allocs/op
func FromCodon(codon string) (string, error) {
	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	}
	return "", ErrInvalidBase
}

File: solutions/ruby/protein-translation/2/protein_translation.rb
=begin
Write your code for the 'Protein Translation' exercise in this file. Make the tests in
`protein_translation_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/protein-translation` directory.
=end

class InvalidCodonError < StandardError
end

class Translation

  CODON_PROTEIN_MAP = {
    AUG: 	"Methionine",
UUU: 	"Phenylalanine",
UUC: 	"Phenylalanine",
UUA: 	"Leucine",
UUG: 	"Leucine",   
UCU: 	"Serine",
UCC: "Serine",
UCA: "Serine",
UCG: "Serine",    
UAU: 	"Tyrosine",
UAC: 	"Tyrosine",    
UGU:"Cysteine",
UGC:"Cysteine",    
UGG:"Tryptophan",
UAA:"STOP",
UAG:"STOP",
UGA:"STOP"    
  }


  
  def self.of_codon(codon)
    CODON_PROTEIN_MAP[codon.to_sym]
  end

  def self.of_rna(strand)
    codons = strand.chars.each_slice(3).map {|item|
      item.join }

    codons.each_with_object([]) {|i, a|
      codon = self.of_codon(i)
    raise InvalidCodonError if codon == nil
      
    return a if codon == "STOP" 
    a.push(codon)
    } 
  end
end